hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/delegation/AbstractDelegationTokenSecretManager.java

```diff
@@ -726,4 +732,41 @@ public TokenIdent decodeTokenIdentifier(Token<TokenIdent> token) throws IOExcept
     return token.decodeIdentifier();
   }
 
+  /**
+   * Return top token real owners list as well as the tokens count.
+   *
+   * @param n top number of users
+   * @return map of owners to counts
+   */
+  public List<NameValuePair> getTopTokenRealOwners(int n) {
+    Map<String, Integer> tokenOwnerMap = new HashMap<>();
+    for (TokenIdent id : currentTokens.keySet()) {
+      String realUser;
+      if (id.getRealUser() != null && !id.getRealUser().toString().isEmpty()) {
+        realUser = id.getRealUser().toString();
+      } else {
+        // if there is no real user -> this is a non proxy user
+        // the user itself is the real owner
+        realUser = id.getUser().getUserName();
+      }
+      tokenOwnerMap.put(realUser, tokenOwnerMap.getOrDefault(realUser, 0)+1);
+    }
+    n = Math.min(n, tokenOwnerMap.size());
+    if (n == 0) {
+      return new LinkedList<>();
+    }
+
+    TopN topN = new TopN(n);
+    for (Map.Entry<String, Integer> entry : tokenOwnerMap.entrySet()) {
+      topN.offer(new NameValuePair(
+          entry.getKey(), entry.getValue()));
+    }
+
+    List<NameValuePair> list = new LinkedList<>();
```

Comment:
any reason to use `LinkedList` instead of `ArrayList`? the latter is usually more performant.

Reply:
There is a reverse op and I think reverse linked list is faster without additional space.
Not sure how java implement the reverse array list, but I think it will introduce copy and reassign.

Reply:
Reverse shouldn't need extra space - it uses two indexes from begin and end of the array and swaps elements. I don't see real difference between the two for the reverse.